### PR TITLE
feat: support truffleruby

### PR DESCRIPTION
**Pull request by ahayworth**
Does what it says on the tin. By "support" I mean "turn it on in CI and
see if everything works."

The compatibility list is actually a bit better than JRuby!

Notably, I needed to exclude the OTLP exporter to make tests pass. I'm
not certain it's a real bug; I just haven't tried to fix those test
failures yet.

---

For discussion: Do we actually even want this? I went down a rabbithole today and landed on truffleruby, and thought it might be neat to support it. It seems like it's broadly more compatible than JRuby to me. Digging through past issues and PRs, it looks like we've at least attempted to support this in the past, and then maybe accidentally dropped it over time.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -306,13 +350,30 @@ jobs:
...
         uses: ./.github/actions/test_gem
         with:
           gem: "${{ matrix.gem }}"
           ruby: "jruby"
+      - name: "truffleruby Filter"
+        id: truffleruby_skip
+        shell: bash
+        run: |
+          echo "::set-output name=skip::false"
+          [[ "${{ matrix.gem }}" == "opentelemetry-instrumentation-que"        ]] && echo "::set-output name=skip::true"
```
Comment (plantfansam) [inline .github/workflows/ci.yml:364]: So these gems fail against truffleruby, is that right?

**Review by plantfansam — APPROVED** [review]
LGTM! If the gems enumerated in `truffleruby filter` don't pass when run in CI, what do you think about making GH issues for each of them? These could make great tasks for newcomers to the project. I am happy to make those tickets if you think it's a good idea.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ahayworth) [reply .github/workflows/ci.yml:364]: Indeed.

**Comment by ahayworth** [comment]
@plantfansam That's a great idea - and perhaps doing it for `jruby` too is also a good idea... I'm not sure how much effort the different gems would be, though.

**Comment by plantfansam** [comment]
> @plantfansam That's a great idea - and perhaps doing it for `jruby` too is also a good idea... I'm not sure how much effort the different gems would be, though.

SGTM. Will make tickets for truffleruby and jruby

**Review by fbogsany — APPROVED** [review]

**Comment by plantfansam** [comment]
Hi @ahayworth! Sorry for the trouble, but with the `opentelemetry-ruby-contrib` split, I think we'll need to chop this PR up into contrib-relevant and this-repo-relevant pieces. There will probably be a merge conflict once https://github.com/open-telemetry/opentelemetry-ruby/pull/1319 makes it into `main`.

**Comment by ahayworth** [comment]
> Hi @ahayworth! Sorry for the trouble, but with the opentelemetry-ruby-contrib split, I think we'll need to chop this PR up into contrib-relevant and this-repo-relevant pieces. There will probably be a merge conflict once https://github.com/open-telemetry/opentelemetry-ruby/pull/1319 makes it into main.

It's no trouble at all; I'll update this.

**Pull request merged**